File: src/dorkbox/cabParser/CabParser.java
/*
 * Copyright 2012 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.cabParser;

import dorkbox.cabParser.decompress.CabDecompressor;
import dorkbox.cabParser.structure.CabEnumerator;
import dorkbox.cabParser.structure.CabFileEntry;
import dorkbox.cabParser.structure.CabFolderEntry;
import dorkbox.cabParser.structure.CabHeader;
import dorkbox.util.process.NullOutputStream;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Enumeration;

public final class CabParser {
    private CabInputStream cabInputStream;

    private CabStreamSaver streamSaver;
    private ByteArrayOutputStream outputStream = null;

    public CabHeader header;

    public CabFolderEntry[] folders;
    public CabFileEntry[]   files;

    public
    CabParser(InputStream inputStream, final String fileNameToExtract) throws CabException, IOException {
        if (fileNameToExtract == null || fileNameToExtract.isEmpty()) {
            throw new IllegalArgumentException("Filename must be valid!");
        }

        this.cabInputStream = new CabInputStream(inputStream);
        this.streamSaver = new CabStreamSaver() {
            @Override
            public boolean saveReservedAreaData(byte[] data, int dataLength) {
                return false;
            }

            @Override
            public OutputStream openOutputStream(CabFileEntry cabFile) {
                String name = cabFile.getName();
                if (fileNameToExtract.equalsIgnoreCase(name)) {
                    CabParser.this.outputStream = new ByteArrayOutputStream((int) cabFile.getSize());
                    return CabParser.this.outputStream;
                } else {
                    return null;
                }
            }

            @Override
            public void closeOutputStream(OutputStream outputStream, CabFileEntry cabFile) {
                if (outputStream != null) {
                    try {
                        outputStream.close();
                    } catch (IOException ignored) {
                    }
                }
            }
        };

        readData();
    }

    public
    CabParser(InputStream inputStream, CabStreamSaver streamSaver) throws CabException, IOException {
        this.streamSaver = streamSaver;
        this.cabInputStream = new CabInputStream(inputStream);

        readData();
    }

    /**
     * Gets the version number.
     */
    public static
    String getVersion() {
        return "2.9";
    }

    public Enumeration<Object> entries() {
        return new CabEnumerator(this, false);
    }

    public Enumeration<Object> entries(boolean b) {
        return new CabEnumerator(this, b);
    }

    private void readData() throws CabException, IOException {
        this.header = new CabHeader(this.streamSaver);
        this.header.read(this.cabInputStream);

        if (this.header.cbCabinet <= 2147483647L) {
            this.cabInputStream.mark((int) this.header.cbCabinet);
        }

        this.folders = new CabFolderEntry[this.header.cFolders];
        for (int i = 0; i < this.header.cFolders; i++) {
            this.folders[i] = new CabFolderEntry();
            this.folders[i].read(this.cabInputStream);
        }

        this.files = new CabFileEntry[this.header.cFiles];
        this.cabInputStream.seek(this.header.coffFiles);

        for (int i = 0; i < this.header.cFiles; i++) {
            this.files[i] = new CabFileEntry();
            this.files[i].read(this.cabInputStream);
        }

        if (this.header.cbCabinet <= 2147483647L) {
            this.cabInputStream.mark((int) this.header.cbCabinet);
        }
    }

    public ByteArrayOutputStream extractStream() throws CabException, IOException {
        int folderCount = -1;
        int currentCount = 0;
        int totalCount = 0;
        boolean init = true;

        CabDecompressor extractor = new CabDecompressor(this.cabInputStream, this.header.cbCFData);

        for (int fileIndex = 0; fileIndex < this.header.cFiles; fileIndex++) {
            CabFileEntry entry = this.files[fileIndex];

            if (entry.iFolder != folderCount) {
                if (folderCount + 1 >= this.header.cFolders) {
                    throw new CorruptCabException();
                }

                folderCount++;
                init = true;
                currentCount = 0;
                totalCount = 0;
            }

            OutputStream localOutputStream = this.streamSaver.openOutputStream(entry);
            if (localOutputStream != null) {
                if (init) {
                    CabFolderEntry cabFolderEntry = this.folders[folderCount];

                    this.cabInputStream.seek(cabFolderEntry.coffCabStart);
                    extractor.initialize(cabFolderEntry.compressionMethod);
                    init = false;
                }

                if (currentCount != totalCount) {
                    extractor.read(totalCount - currentCount, new NullOutputStream());
                    currentCount = totalCount;
                }

                extractor.read(entry.cbFile, localOutputStream);
                this.streamSaver.closeOutputStream(localOutputStream, entry);
                currentCount = (int) (currentCount + entry.cbFile);
            }

            totalCount = (int) (totalCount + entry.cbFile);
        }

        return this.outputStream;
    }
}

File: src/dorkbox/cabParser/CabParser.kt
/*
 * Copyright 2023 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.cabParser

import dorkbox.cabParser.decompress.CabDecompressor
import dorkbox.cabParser.structure.CabEnumerator
import dorkbox.cabParser.structure.CabFileEntry
import dorkbox.cabParser.structure.CabFolderEntry
import dorkbox.cabParser.structure.CabHeader
import dorkbox.updates.Updates.add
import java.io.*
import java.util.*

class CabParser {
    companion object {
        /**
         * Gets the version number.
         */
        const val version: String = "3.2"

        init {
            // Add this project to the updates system, which verifies this class + UUID + version information
            add(CabParser::class.java, "41f560ca51c04bfdbca21328e0cbf206", version)
        }
    }

    private var cabInputStream: CabInputStream
    private var streamSaver: CabStreamSaver
    private var outputStream: ByteArrayOutputStream? = null

    lateinit var header: CabHeader

    lateinit var folders: Array<CabFolderEntry>

    lateinit var files: Array<CabFileEntry>

    constructor(inputStream: InputStream, fileNameToExtract: String) {
        cabInputStream = CabInputStream(inputStream)
        streamSaver = object : CabStreamSaver {
            override fun saveReservedAreaData(data: ByteArray?, dataLength: Int): Boolean {
                return false
            }

            override fun openOutputStream(entry: CabFileEntry): OutputStream? {
                val name = entry.name
                return if (fileNameToExtract.equals(name, ignoreCase = true)) {
                    outputStream = ByteArrayOutputStream(entry.size.toInt())
                    outputStream!!
                }
                else {
                    null
                }
            }

            override fun closeOutputStream(outputStream: OutputStream, cabFile: CabFileEntry) {
                try {
                    outputStream.close()
                }
                catch (ignored: IOException) {
                }
            }
        }
        readData()
    }

    constructor(inputStream: InputStream, streamSaver: CabStreamSaver) {
        this.streamSaver = streamSaver
        cabInputStream = CabInputStream(inputStream)
        readData()
    }

    constructor(inputStream: InputStream, extractPath: File?) {
        streamSaver = DefaultCabStreamSaver(extractPath)
        cabInputStream = CabInputStream(inputStream)
        readData()
    }

    fun entries(): Enumeration<Any> {
        return CabEnumerator(this, false)
    }

    fun entries(b: Boolean): Enumeration<Any> {
        return CabEnumerator(this, b)
    }

    @Suppress("UNCHECKED_CAST")
    @Throws(CabException::class, IOException::class)
    private fun readData() {
        header = CabHeader(streamSaver)
        header.read(cabInputStream)
        folders = arrayOfNulls<CabFolderEntry>(header.cFolders) as Array<CabFolderEntry>

        for (i in 0 until header.cFolders) {
            folders[i] = CabFolderEntry()
            folders[i].read(cabInputStream)
        }

        files = arrayOfNulls<CabFileEntry>(header.cFiles) as Array<CabFileEntry>
        cabInputStream.seek(header.coffFiles)
        for (i in 0 until header.cFiles) {
            files[i] = CabFileEntry()
            files[i].read(cabInputStream)
        }
    }

    @Throws(CabException::class, IOException::class)
    fun extractStream(): ByteArrayOutputStream? {
        var folderCount = -1
        var currentCount = 0
        var totalCount = 0
        var init = true
        val extractor = CabDecompressor(cabInputStream, header.cbCFData)
        for (fileIndex in 0 until header.cFiles) {
            val entry = files[fileIndex]
            if (entry.iFolder != folderCount) {
                if (folderCount + 1 >= header.cFolders) {
                    throw CorruptCabException()
                }

                folderCount++
                init = true
                currentCount = 0
                totalCount = 0
            }
            val localOutputStream = streamSaver.openOutputStream(entry)
            if (localOutputStream != null) {
                if (init) {
                    val cabFolderEntry = folders[folderCount]
                    cabInputStream.seek(cabFolderEntry.coffCabStart)
                    extractor.initialize(cabFolderEntry.compressionMethod)
                    init = false
                }
                if (currentCount != totalCount) {
                    extractor.read((totalCount - currentCount).toLong(), object : OutputStream() {
                        @Throws(IOException::class)
                        override fun write(i: Int) {
                            //do nothing
                        }

                        @Throws(IOException::class)
                        override fun write(b: ByteArray) {
                            //do nothing
                        }

                        @Throws(IOException::class)
                        override fun write(b: ByteArray, off: Int, len: Int) {
                            //do nothing
                        }

                        @Throws(IOException::class)
                        override fun flush() {
                            //do nothing
                        }
                    })
                    currentCount = totalCount
                }
                extractor.read(entry.cbFile, localOutputStream)
                streamSaver.closeOutputStream(localOutputStream, entry)
                currentCount = (currentCount + entry.cbFile).toInt()
            }
            totalCount = (totalCount + entry.cbFile).toInt()
        }
        return outputStream
    }
}
